js/publisher.js

```diff
@@ -53,80 +45,126 @@ class JS extends Publisher {
+      this.initSynopsis()
     }
 
+    this.setState(this.setLocationData(publisher))
+    this.setState(this.ledger.addNewLocation(this.state, publisher.url, publisher.tabId, false, true))
     this.setState(this.ledger.pageDataChanged(this.state, {
-      location: publisherUrl,
-      tabId: publisherTabId
+      location: publisher.url,
+      tabId: publisher.tabId
     }))
+  }
 
-    return this.ledger.getSynopsis()
+  addPublisherVisit (publisher, enableSynopsis = true) {
+    if (enableSynopsis) {
+      this.initSynopsis()
+    }
+
+    this.currentVisitTime = publisher.visitTime
+    this.setState(this.setLocationData(publisher))
+    this.setState(this.ledger.pageDataChanged(this.state, {
+      location: publisher.url,
+      tabId: publisher.tabId
+    }))
+    this.setState(this.ledger.pageDataChanged(this.state, {}, true))
   }
 
-  get mediaRequest () {
-    return responses['media'][this.mediaType]['media-request'][this.mediaMinimum]
+  deletePublisher (publisherKey) {
+    this.ledger.deleteSynopsisPublisher(publisherKey)
+    this.setState(ledgerState.deletePublishers(this.state, publisherKey))
+    this.setState(this.ledger.updatePublisherInfo(this.state, publisherKey))
   }
 
   pinPublisher (publisherKey, percentage, pinned = true) {
+    const newPercentage = pinned ? percentage : 0
     const publisher = ledgerState.getPublisher(this.state, publisherKey)
 
     if (publisher.isEmpty()) {
       return
     }
 
     this.setState(this.ledger.updatePublisherInfo(this.state))
+    this.setState(ledgerState.setPublishersProp(this.state, publisherKey, 'pinPercentage', newPercentage))
 
-    this.setState(ledgerState.setPublishersProp(this.state, publisherKey, 'pinPercentage', percentage))
-    this.ledger.savePublisherData(publisherKey, 'pinPercentage', percentage)
+    this.ledger.savePublisherData(publisherKey, 'pinPercentage', newPercentage)
     this.setState(this.ledger.updatePublisherInfo(this.state, publisherKey))
+  }
 
-    return this.ledger.getSynopsis()
+  processTwitchData (key, xhr, type, details) {
+    const twitchUploadData = require('../lib/data/twitchUploadData.json').upload_data[key]
+
+    twitchUploadData.forEach((piece) => {
+      let uploadData = []
+
+      if (!Array.isArray(piece)) {
+        piece.data.forEach((data) => {
+          uploadData.push({
+            'bytes': Buffer.from(data)
+          })
+        })
+      } else {
+        uploadData = [{'bytes': Buffer.from(piece)}]
+      }
+
+      details.uploadData = uploadData
+      this.setState(this.ledger.onMediaRequest(this.state, xhr, type, Immutable.fromJS(details)))
+    })
   }
 
-  invokeMediaRequest (type, min = false) {
+  invokeMediaRequest (type, min = false, key = false) {
     this.mediaType = type
     this.mediaMinimum = min ? 'min' : 'non'
     this.state = this.ledger.enable(this.defaultAppState)
 
     const xhr = this.mediaRequest.xhr
     const details = Immutable.fromJS(this.mediaRequest.details)
```

Comment:
Can we rename this one into detailImmutable and have regular details as well. This way you don't need to do toJS in line 163